pyacq/core/stream.py

```diff
@@ -285,84 +284,157 @@ class PlainDataSender:
+        index = stat[0]
+        offset = stat[1]
+        shape = stat[2:2+ndim]
+        strides = stat[-ndim:]
+        
+        # uncompress
+        comp = self.params['compression']
+        if comp.startswith('blosc-'):
+            assert HAVE_BLOSC, "Cannot use blosc decompression; blosc package is not importable."
+            data = blosc.decompress(data)
+        elif comp != '':
+            raise ValueError("Unknown compression method '%s'" % comp)
+        
+        # convert to array
+        dtype = self.params['dtype']
+        data = np.ndarray(buffer=data, shape=shape,
+                          strides=strides, offset=offset, dtype=dtype)        
+        return index, data
+
+
+class SharedMemSender:
+    """
+    """
+    def __init__(self, socket, params):
         self.socket = socket
         self.params = params
-        
-        if self.params['compression'] == '':
-            self.func = self._numpy_fromstring
-        elif self.params['compression'] in ['blosc-blosclz', 'blosc-lz4']:
-            self.func = self._uncompress_blosc
+        self.size = self.params['buffer_size']
+        dtype = np.dtype(self.params['dtype'])
+        shape = (self.size,) + self.params['shape'][1:]
+        self._buffer = RingBuffer(shape=shape, dtype=self.params['dtype'],
+                                  shmem=True, axisorder=self.params['axisorder'])
+        self.params['shm_id'] = self._buffer.shm_id
     
-    def recv(self, autoswapaxes = True):
-        m0,m1 = self.socket.recv_multipart()
-        index = np.fromstring(m0, dtype='int64')[0]
-        index, data = self.func(index, m1)
-        if autoswapaxes and self.params['timeaxis']!=0:
-            data = data.swapaxes(0, self.params['timeaxis'])
-        return  index, data
+    def send(self, index, data):
+        assert data.dtype == self.params['dtype']
+        shape = data.shape
+        if self.params['shape'][0] != -1:
+            assert shape == self.params['shape']
+        else:
+            assert shape[1:] == self.params['shape'][1:]
+ 
+        self._buffer.new_chunk(data, index)
+        
+        stat = struct.pack('!' + 'Q' * (2+len(shape)), len(shape), index, *shape)
+        self.socket.send_multipart([stat])
     
     def close(self):
         pass
 
-    def _numpy_fromstring(self, index, data):
-        data = np.frombuffer(data, dtype=self.params['dtype']).reshape(self.params['shape'])
+
+class SharedMemReceiver(DataReceiver):
+    def __init__(self, socket, params):
+        # init data receiver with no ring buffer; we will implement our own from shm.
+        DataReceiver.__init__(self, socket, params)
+
+        self.size = self.params['buffer_size']
+        shape = (self.size,) + self.params['shape'][1:]
+        self._buffer = RingBuffer(shape=shape, dtype=self.params['dtype'],
+                                  shmem=self.params['shm_id'], axisorder=self.params['axisorder'])
+
+    def _recv(self):
```

Comment:
In the old behavior for sharedarrayreceiver, the data was always None and only the index had sens.

The reasons was : in sharredarray buffer, often the last chunk have no meaning computations are done on longer chunk sliding chunk. So we need a case where data is not constructed at each send/recv. I guess have no proof that intensive send/recv this may count to not construct last data chunk.
So I propose a kargs in init: with_data=True/False and with False 
```data = self._buffer[index+1-shape[0]:index+1]``` would be skipped.